Cabal >= 3 does not pass through options to $CC correctly when building C code for FFI modules.

A project built with `llvm-hs` gives an error like this in GHC 8.10.3:
```
Linking .stack-work/dist/x86_64-linux-tinfo6/Cabal-3.2.1.0/build/<MyProject>/<MyProject> ...
ld.lld: error: undefined symbol: typeinfo for llvm::ErrorInfoBase
>>> referenced by BitcodeC.o:(typeinfo for llvm::ErrorInfo<llvm::ErrorList, llvm::ErrorInfoBase>) in archive <HOME>/.stack/snapshots/x86_64-linux-tinfo6/d1a1b99df71b047a18276db4b4d4dcaa7c67873cb82e9c6788ea3b790131ffee/8.10.3/lib/x86_64-linux-ghc-8.10.3/llvm-hs-9.0.1-1RtPvEeO6wN3mXnaO6NE50/libHSllvm-hs-9.0.1-1RtPvEeO6wN3mXnaO6NE50.a
collect2: error: ld returned 1 exit status
`gcc' failed in phase `Linker'. (Exit code: 1)
Completed 14 action(s).             

--  While building package <MyProject>-0.1.0.0 (scroll up to its section to see the error) using:
      <HOME>/.stack/setup-exe-cache/x86_64-linux-tinfo6/Cabal-simple_mPHDZzAJ_3.2.1.0_ghc-8.10.3 --builddir=.stack-work/dist/x86_64-linux-tinfo6/Cabal-3.2.1.0 build exe:<MyProject> --ghc-options " -fdiagnostics-color=always"
    Process exited with code: ExitFailure 1
```
, though I'm not sure it is generally reproducible with other GHC 8.10 versions.

Is this expected?

(FYI, it works well with GHC 8.8.4)

```
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1/In file included from src/LLVM/Internal/FFI/AssemblyC.cpp:2:0: error:
llvm-hs             >     
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1/In file included from /usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/AsmParser/Parser.h:16:0: error:
llvm-hs             >     
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1/In file included from /usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/MemoryBuffer.h:17:0: error:
llvm-hs             >     
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1/In file included from /usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/ADT/ArrayRef.h:12:0: error:
llvm-hs             >     
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1/In file included from /usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/ADT/Hashing.h:48:0: error:
llvm-hs             >     
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1/In file included from /usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/Host.h:16:0: error:
llvm-hs             >     
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1/In file included from /usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/ADT/StringMap.h:16:0: error:
llvm-hs             >     
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1/In file included from /usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/ADT/StringRef.h:12:0: error:
llvm-hs             >     
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1/In file included from /usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/ADT/STLExtras.h:19:0: error:
llvm-hs             >     
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1/In file included from /usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/ADT/Optional.h:18:0: error:
llvm-hs             >     
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/ADT/None.h:22:6: error:
llvm-hs             >      warning: scoped enumerations are a C++11 extension [-Wc++11-extensions]
llvm-hs             >    |
llvm-hs             > 22 | enum class NoneType { None = 1 };
llvm-hs             >    |      ^
llvm-hs             > enum class NoneType { None = 1 };
llvm-hs             >      ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/ADT/None.h:23:23: error:
llvm-hs             >      warning: use of enumeration in a nested name specifier is a C++11 extension [-Wc++11-extensions]
llvm-hs             >    |
llvm-hs             > 23 | const NoneType None = NoneType::None;
llvm-hs             >    |                       ^
llvm-hs             > const NoneType None = NoneType::None;
llvm-hs             >                       ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1/In file included from src/LLVM/Internal/FFI/AssemblyC.cpp:2:0: error:
llvm-hs             >     
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1/In file included from /usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/AsmParser/Parser.h:16:0: error:
llvm-hs             >     
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1/In file included from /usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/MemoryBuffer.h:17:0: error:
llvm-hs             >     
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1/In file included from /usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/ADT/ArrayRef.h:12:0: error:
llvm-hs             >     
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1/In file included from /usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/ADT/Hashing.h:48:0: error:
llvm-hs             >     
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1/In file included from /usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/Host.h:16:0: error:
llvm-hs             >     
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1/In file included from /usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/ADT/StringMap.h:16:0: error:
llvm-hs             >     
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1/In file included from /usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/ADT/StringRef.h:12:0: error:
llvm-hs             >     
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1/In file included from /usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/ADT/STLExtras.h:19:0: error:
llvm-hs             >     
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1/In file included from /usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/ADT/Optional.h:20:0: error:
llvm-hs             >     
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:36:23: error:
llvm-hs             >      warning: alias declarations are a C++11 extension [-Wc++11-extensions]
llvm-hs             >       using UnderlyingT = typename std::remove_reference<T>::type;
llvm-hs             >                           ^
llvm-hs             >    |
llvm-hs             > 36 |   using UnderlyingT = typename std::remove_reference<T>::type;
llvm-hs             >    |                       ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:49:59: error:
llvm-hs             >      warning: alias declarations are a C++11 extension [-Wc++11-extensions]
llvm-hs             >    |
llvm-hs             > 49 | struct add_lvalue_reference_if_not_pointer { using type = T &; };
llvm-hs             >    |                                                           ^
llvm-hs             > struct add_lvalue_reference_if_not_pointer { using type = T &; };
llvm-hs             >                                                           ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:54:16: error:
llvm-hs             >      warning: alias declarations are a C++11 extension [-Wc++11-extensions]
llvm-hs             >       using type = T;
llvm-hs             >                    ^
llvm-hs             >    |
llvm-hs             > 54 |   using type = T;
llvm-hs             >    |                ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:60:46: error:
llvm-hs             >      warning: alias declarations are a C++11 extension [-Wc++11-extensions]
llvm-hs             >    |
llvm-hs             > 60 | struct add_const_past_pointer { using type = const T; };
llvm-hs             >    |                                              ^
llvm-hs             > struct add_const_past_pointer { using type = const T; };
llvm-hs             >                                              ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:65:16: error:
llvm-hs             >      warning: alias declarations are a C++11 extension [-Wc++11-extensions]
llvm-hs             >       using type = const typename std::remove_pointer<T>::type *;
llvm-hs             >                    ^
llvm-hs             >    |
llvm-hs             > 65 |   using type = const typename std::remove_pointer<T>::type *;
llvm-hs             >    |                ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:70:16: error:
llvm-hs             >      warning: alias declarations are a C++11 extension [-Wc++11-extensions]
llvm-hs             >       using type = const T &;
llvm-hs             >                    ^
llvm-hs             >    |
llvm-hs             > 70 |   using type = const T &;
llvm-hs             >    |                ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:75:16: error:
llvm-hs             >      warning: alias declarations are a C++11 extension [-Wc++11-extensions]
llvm-hs             >       using type = typename add_const_past_pointer<T>::type;
llvm-hs             >                    ^
llvm-hs             >    |
llvm-hs             > 75 |   using type = typename add_const_past_pointer<T>::type;
llvm-hs             >    |                ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:82:45: error:
llvm-hs             >      warning: defaulted function definitions are a C++11 extension [-Wc++11-extensions]
llvm-hs             >         copy_construction_triviality_helper() = default;
llvm-hs             >                                                 ^
llvm-hs             >    |
llvm-hs             > 82 |     copy_construction_triviality_helper() = default;
llvm-hs             >    |                                             ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:83:87: error:
llvm-hs             >      warning: defaulted function definitions are a C++11 extension [-Wc++11-extensions]
llvm-hs             >         copy_construction_triviality_helper(const copy_construction_triviality_helper&) = default;
llvm-hs             >                                                                                           ^
llvm-hs             >    |
llvm-hs             > 83 |     copy_construction_triviality_helper(const copy_construction_triviality_helper&) = default;
llvm-hs             >    |                                                                                       ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:84:46: error:
llvm-hs             >      warning: defaulted function definitions are a C++11 extension [-Wc++11-extensions]
llvm-hs             >         ~copy_construction_triviality_helper() = default;
llvm-hs             >                                                  ^
llvm-hs             >    |
llvm-hs             > 84 |     ~copy_construction_triviality_helper() = default;
llvm-hs             >    |                                              ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:89:45: error:
llvm-hs             >      warning: defaulted function definitions are a C++11 extension [-Wc++11-extensions]
llvm-hs             >         move_construction_triviality_helper() = default;
llvm-hs             >                                                 ^
llvm-hs             >    |
llvm-hs             > 89 |     move_construction_triviality_helper() = default;
llvm-hs             >    |                                             ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:90:76: error:
llvm-hs             >      warning: rvalue references are a C++11 extension [-Wc++11-extensions]
llvm-hs             >         move_construction_triviality_helper(move_construction_triviality_helper&&) = default;
llvm-hs             >                                                                                ^
llvm-hs             >    |
llvm-hs             > 90 |     move_construction_triviality_helper(move_construction_triviality_helper&&) = default;
llvm-hs             >    |                                                                            ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:90:82: error:
llvm-hs             >      warning: defaulted function definitions are a C++11 extension [-Wc++11-extensions]
llvm-hs             >         move_construction_triviality_helper(move_construction_triviality_helper&&) = default;
llvm-hs             >                                                                                      ^
llvm-hs             >    |
llvm-hs             > 90 |     move_construction_triviality_helper(move_construction_triviality_helper&&) = default;
llvm-hs             >    |                                                                                  ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:91:46: error:
llvm-hs             >      warning: defaulted function definitions are a C++11 extension [-Wc++11-extensions]
llvm-hs             >         ~move_construction_triviality_helper() = default;
llvm-hs             >                                                  ^
llvm-hs             >    |
llvm-hs             > 91 |     ~move_construction_triviality_helper() = default;
llvm-hs             >    |                                              ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:106:64: error:
llvm-hs             >      error: a space is required between consecutive right angle brackets (use '> >')
llvm-hs             >               ::llvm::detail::copy_construction_triviality_helper<T>> {};
llvm-hs             >                                                                    ^~
llvm-hs             >                                                                    > >
llvm-hs             >     |
llvm-hs             > 106 |           ::llvm::detail::copy_construction_triviality_helper<T>> {};
llvm-hs             >     |                                                                ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:110:42: error:
llvm-hs             >      warning: rvalue references are a C++11 extension [-Wc++11-extensions]
llvm-hs             >     |
llvm-hs             > 110 | struct is_trivially_copy_constructible<T &&> : std::false_type {};
llvm-hs             >     |                                          ^
llvm-hs             > struct is_trivially_copy_constructible<T &&> : std::false_type {};
llvm-hs             >                                          ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:117:64: error:
llvm-hs             >      error: a space is required between consecutive right angle brackets (use '> >')
llvm-hs             >               ::llvm::detail::move_construction_triviality_helper<T>> {};
llvm-hs             >                                                                    ^~
llvm-hs             >                                                                    > >
llvm-hs             >     |
llvm-hs             > 117 |           ::llvm::detail::move_construction_triviality_helper<T>> {};
llvm-hs             >     |                                                                ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:121:42: error:
llvm-hs             >      warning: rvalue references are a C++11 extension [-Wc++11-extensions]
llvm-hs             >     |
llvm-hs             > 121 | struct is_trivially_move_constructible<T &&> : std::true_type {};
llvm-hs             >     |                                          ^
llvm-hs             > struct is_trivially_move_constructible<T &&> : std::true_type {};
llvm-hs             >                                          ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:127:12: error:
llvm-hs             >      warning: 'auto' type specifier is a C++11 extension [-Wc++11-extensions]
llvm-hs             >         static auto get(F*) -> decltype(std::declval<F &>() = std::declval<const F &>(), std::true_type{});
llvm-hs             >                ^
llvm-hs             >     |
llvm-hs             > 127 |     static auto get(F*) -> decltype(std::declval<F &>() = std::declval<const F &>(), std::true_type{});
llvm-hs             >     |            ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:127:12: error:
llvm-hs             >      error: 'auto' not allowed in function return type
llvm-hs             >         static auto get(F*) -> decltype(std::declval<F &>() = std::declval<const F &>(), std::true_type{});
llvm-hs             >                ^~~~
llvm-hs             >     |
llvm-hs             > 127 |     static auto get(F*) -> decltype(std::declval<F &>() = std::declval<const F &>(), std::true_type{});
llvm-hs             >     |            ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:127:24: error:
llvm-hs             >      error: expected ';' at end of declaration list
llvm-hs             >         static auto get(F*) -> decltype(std::declval<F &>() = std::declval<const F &>(), std::true_type{});
llvm-hs             >                            ^
llvm-hs             >                            ;
llvm-hs             >     |
llvm-hs             > 127 |     static auto get(F*) -> decltype(std::declval<F &>() = std::declval<const F &>(), std::true_type{});
llvm-hs             >     |                        ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:129:12: error:
llvm-hs             >      error: unknown type name 'constexpr'
llvm-hs             >         static constexpr bool value = decltype(get((T*)nullptr))::value;
llvm-hs             >                ^
llvm-hs             >     |
llvm-hs             > 129 |     static constexpr bool value = decltype(get((T*)nullptr))::value;
llvm-hs             >     |            ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:135:12: error:
llvm-hs             >      warning: 'auto' type specifier is a C++11 extension [-Wc++11-extensions]
llvm-hs             >         static auto get(F*) -> decltype(std::declval<F &>() = std::declval<F &&>(), std::true_type{});
llvm-hs             >                ^
llvm-hs             >     |
llvm-hs             > 135 |     static auto get(F*) -> decltype(std::declval<F &>() = std::declval<F &&>(), std::true_type{});
llvm-hs             >     |            ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:135:12: error:
llvm-hs             >      error: 'auto' not allowed in function return type
llvm-hs             >         static auto get(F*) -> decltype(std::declval<F &>() = std::declval<F &&>(), std::true_type{});
llvm-hs             >                ^~~~
llvm-hs             >     |
llvm-hs             > 135 |     static auto get(F*) -> decltype(std::declval<F &>() = std::declval<F &&>(), std::true_type{});
llvm-hs             >     |            ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:135:24: error:
llvm-hs             >      error: expected ';' at end of declaration list
llvm-hs             >         static auto get(F*) -> decltype(std::declval<F &>() = std::declval<F &&>(), std::true_type{});
llvm-hs             >                            ^
llvm-hs             >                            ;
llvm-hs             >     |
llvm-hs             > 135 |     static auto get(F*) -> decltype(std::declval<F &>() = std::declval<F &&>(), std::true_type{});
llvm-hs             >     |                        ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:137:12: error:
llvm-hs             >      error: unknown type name 'constexpr'
llvm-hs             >         static constexpr bool value = decltype(get((T*)nullptr))::value;
llvm-hs             >                ^
llvm-hs             >     |
llvm-hs             > 137 |     static constexpr bool value = decltype(get((T*)nullptr))::value;
llvm-hs             >     |            ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:149:10: error:
llvm-hs             >      error: unknown type name 'constexpr'
llvm-hs             >       static constexpr bool has_trivial_copy_constructor =
llvm-hs             >              ^
llvm-hs             >     |
llvm-hs             > 149 |   static constexpr bool has_trivial_copy_constructor =
llvm-hs             >     |          ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:150:58: error:
llvm-hs             >      error: a space is required between consecutive right angle brackets (use '> >')
llvm-hs             >           std::is_copy_constructible<detail::trivial_helper<T>>::value;
llvm-hs             >                                                              ^~
llvm-hs             >                                                              > >
llvm-hs             >     |
llvm-hs             > 150 |       std::is_copy_constructible<detail::trivial_helper<T>>::value;
llvm-hs             >     |                                                          ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:151:10: error:
llvm-hs             >      error: unknown type name 'constexpr'
llvm-hs             >       static constexpr bool has_deleted_copy_constructor =
llvm-hs             >              ^
llvm-hs             >     |
llvm-hs             > 151 |   static constexpr bool has_deleted_copy_constructor =
llvm-hs             >     |          ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:155:10: error:
llvm-hs             >      error: unknown type name 'constexpr'
llvm-hs             >       static constexpr bool has_trivial_move_constructor =
llvm-hs             >              ^
llvm-hs             >     |
llvm-hs             > 155 |   static constexpr bool has_trivial_move_constructor =
llvm-hs             >     |          ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:156:58: error:
llvm-hs             >      error: a space is required between consecutive right angle brackets (use '> >')
llvm-hs             >           std::is_move_constructible<detail::trivial_helper<T>>::value;
llvm-hs             >                                                              ^~
llvm-hs             >                                                              > >
llvm-hs             >     |
llvm-hs             > 156 |       std::is_move_constructible<detail::trivial_helper<T>>::value;
llvm-hs             >     |                                                          ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:157:10: error:
llvm-hs             >      error: unknown type name 'constexpr'
llvm-hs             >       static constexpr bool has_deleted_move_constructor =
llvm-hs             >              ^
llvm-hs             >     |
llvm-hs             > 157 |   static constexpr bool has_deleted_move_constructor =
llvm-hs             >     |          ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:161:10: error:
llvm-hs             >      error: unknown type name 'constexpr'
llvm-hs             >       static constexpr bool has_trivial_copy_assign =
llvm-hs             >              ^
llvm-hs             >     |
llvm-hs             > 161 |   static constexpr bool has_trivial_copy_assign =
llvm-hs             >     |          ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:162:50: error:
llvm-hs             >      error: a space is required between consecutive right angle brackets (use '> >')
llvm-hs             >           is_copy_assignable<detail::trivial_helper<T>>::value;
llvm-hs             >                                                      ^~
llvm-hs             >                                                      > >
llvm-hs             >     |
llvm-hs             > 162 |       is_copy_assignable<detail::trivial_helper<T>>::value;
llvm-hs             >     |                                                  ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:163:10: error:
llvm-hs             >      error: unknown type name 'constexpr'
llvm-hs             >       static constexpr bool has_deleted_copy_assign =
llvm-hs             >              ^
llvm-hs             >     |
llvm-hs             > 163 |   static constexpr bool has_deleted_copy_assign =
llvm-hs             >     |          ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:167:10: error:
llvm-hs             >      error: unknown type name 'constexpr'
llvm-hs             >       static constexpr bool has_trivial_move_assign =
llvm-hs             >              ^
llvm-hs             >     |
llvm-hs             > 167 |   static constexpr bool has_trivial_move_assign =
llvm-hs             >     |          ^
llvm-hs             >  
llvm-hs             > /private/var/folders/qz/7zfds_gs7bj6vh_5hrz747p40000gn/T/stack-f32ef84a9e4d390a/llvm-hs-9.0.1//usr/local/Cellar/llvm-9/9.0.1/lib/llvm-9/include/llvm/Support/type_traits.h:168:50: error:
llvm-hs             >      error: a space is required between consecutive right angle brackets (use '> >')
llvm-hs             >           is_move_assignable<detail::trivial_helper<T>>::value;
llvm-hs             >                                                      ^~
llvm-hs             >                                                      > >
llvm-hs             >     |
llvm-hs             > 168 |       is_move_assignable<detail::trivial_helper<T>>::value;
llvm-hs             >     |                                                  ^
llvm-hs             > fatal error: too many errors emitted, stopping now [-ferror-limit=]
llvm-hs             > 20 warnings and 20 errors generated.
llvm-hs             > `gcc' failed in phase `C Compiler'. (Exit code: 1)

```

--- a/README.md
+++ b/README.md
@@ -25,6 +25,24 @@ We love all kinds of contributions so feel free to open issues for
 missing LLVM features, report & fix bugs or report API
 inconveniences.
 
+## Build Problems with GHC 8.10/Cabal 3
+
+If you are building a project that uses llvm-hs, you may encounter compile-time
+and link-time failures using GHC 8.10 / Cabal 3.0 and later. Since llvm-hs
+binds to the LLVM C API via Haskell's FFI, we must build and link against
+some C++ code. Cabal does not propagate options to the backend C/C++ compiler
+correctly, so you must also pass these options at the level of your client
+project. If you are using stack and you see undefined references to `typeinfo`
+and `vtable`, you need to pass the following options in your `stack.yaml`
+
+```
+ghc-options:
+  llvm-hs: -optcxx=-std=c++14 -optcxx=-lstdc++ -optcxx=-fno-rtti
+```
+
+Similarly, if you are using Cabal directly, you need to pass these options via
+`cc-options` in your project `.cabal` file.
+
 ## Installing LLVM
 
 LLVM 12 is still not fully released, and as such is unavailable in most

--- a/llvm-hs/llvm-hs.cabal
+++ b/llvm-hs/llvm-hs.cabal
@@ -224,7 +224,7 @@ library
     src/LLVM/Internal/FFI/TypeC.cpp
     src/LLVM/Internal/FFI/ValueC.cpp
 
-  cc-options: -std=c++11 -Wno-stringop-overflow -Wno-init-list-lifetime
+  cc-options: -std=c++14 -Wno-stringop-overflow -Wno-init-list-lifetime
 
   if flag(debug)
     cc-options: -g

--- a/llvm-hs/llvm-hs.cabal
+++ b/llvm-hs/llvm-hs.cabal
@@ -224,7 +224,7 @@ library
     src/LLVM/Internal/FFI/TypeC.cpp
     src/LLVM/Internal/FFI/ValueC.cpp
 
-  cc-options: -std=c++14 -Wno-stringop-overflow -Wno-init-list-lifetime
+  cc-options: -std=c++14 -lstdc++ -fno-rtti -Wno-stringop-overflow -Wno-init-list-lifetime
 
   if flag(debug)
     cc-options: -g

--- a/stack.yaml
+++ b/stack.yaml
@@ -1,4 +1,4 @@
-resolver: lts-16.31
+resolver: lts-17.12
 
 packages:
 - llvm-hs

--- a/stack.yaml
+++ b/stack.yaml
@@ -6,3 +6,5 @@ packages:
 
 extra-package-dbs: []
 
+ghc-options:
+  llvm-hs: -optcxx=-std=c++14 -optcxx=-lstdc++ -optcxx=-fno-rtti

